### PR TITLE
Garbage collector for active calls

### PR DESCRIPTION
**Pull request by aeatencio**
When a call remains active for too long Verboice considers there was an error and cancels it.

This GC runs every X minutes and cancels every active call for more than N minutes.

The value X defaults to 10 minutes and is configurable via the ENV variable `minutes_between_active_calls_gc_runs`.

The value N defaults to 120 minutes and is configurable via the ENV variable `minutes_for_cancelling_active_calls`.

For #900

Bonus track: buggy code deletion.

### CODE REVIEW

File: broker/src/models/call_log.erl [hunk 1]
```diff
@@ -12,3 +12,8 @@ info(Message, Details, #call_log{id = CallId}) ->
 trace(Message, Details, #call_log{id = CallId}) ->
   call_log_entry:create("trace", CallId, Message, Details).
 
+cancel_active_calls_for_minutes(N, Reason) ->
+  N_Minutes_Ago = util:seconds_ago(N * 60),
+  Count = call_log:update_all([{state, "failed"}, {fail_reason, Reason}], [{state, "active"}, {started_at, '<', N_Minutes_Ago}]),
```
Comment (matiasgarciaisaia) [inline broker/src/models/call_log.erl:17]: The `fail_reason` field seems to be kind of pre-fixed (they look like symbols, like `no-answer`, `no-call-flow` or `internal error`). Think of it as a field by which you can easily index/search for calls.

And then you can be a bit more verbose on the `fail_details`. I'd set both of them here 👍

File: broker/src/scheduler/active_calls_gc.erl [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+handle_call(_Request, _From, State) ->
+  {reply, {error, unknown_call}, State}.
+
+handle_cast(_Msg, State) ->
+  {noreply, State}.
+
+handle_info(cancel_active_calls, State) ->
+  N = verboice_config:minutes_for_cancelling_active_calls(),
+  Reason = "active_calls_gc N=" ++ integer_to_list(N),
```
Comment (matiasgarciaisaia) [inline broker/src/scheduler/active_calls_gc.erl:33]: I'd rephrase this in a more user-friendly way. `active-too-long` or `kept-active` or `remained-active` or similar, and then add the details like `Call stayed active for more than N minutes`.

File: broker/src/scheduler/active_calls_gc.erl [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  Count = call_log:cancel_active_calls_for_minutes(N, Reason),
+  lager:info("Active calls GC run with N=~p and Count=~p", [N, Count]),
```
Comment (matiasgarciaisaia) [inline broker/src/scheduler/active_calls_gc.erl:35]:
```suggestion
  lager:info("GC cancelled ~p calls that stayed active for more than ~p minutes", [Count, N]),
```

**Review by matiasgarciaisaia — COMMENTED** [review]
Do we know if the broker _understands_ that a call got cancelled with respect to the channel limit? ie, will it start a new call if any one is queued?

I'm also thinking about the callbacks - will they trigger? My fear is that we're silently just marking calls as failed, without that affecting the behaviour of the system.

**Comment by aeatencio** [comment]
> Do we know if the broker understands that a call got cancelled with respect to the channel limit? ie, will it start a new call if any one is queued?

@matiasgarciaisaia, you may be right! Thanks!!!

[This](https://github.com/instedd/verboice/blob/178b27c540fefd369c7fbec264de86947e731609/app/controllers/api/calls_controller.rb#L94-L111) is what we do in the API. Maybe we need to include `queued_calls` [this way](https://github.com/instedd/verboice/blob/7d3e02f4386d0c2b3788ca8d0b40a188f9ecf96a/app/models/queued_call.rb#L35-L38) too, right?

> I'm also thinking about the callbacks - will they trigger? My fear is that we're silently just marking calls as failed, without that affecting the behavior of the system.

No clue about how to check this. But what I think is that we're reproducing how the calls are being canceled manually, so it should work, right?

File: app/controllers/api/calls_controller.rb [hunk 4]
```diff
@@ -95,12 +95,6 @@ def cancel
       call_log = current_account.call_logs.where(:id => params[:id]).first
       return head :not_found if call_log.nil?
 
-      queued_calls = QueuedCall.where(call_log_id: call_log.id).all
-      queued_calls.each do |qc|
-        qc.cancel_call!
-        qc.destroy
```
Comment (matiasgarciaisaia) [inline app/controllers/api/calls_controller.rb:101]: Destroying the queued calls is probably still important/required

**Comment by aeatencio** [comment]
@matiasgarciaisaia, please let me know how you see these changes now. I know this may not be the ideal solution, but please let me know if you think it's good enough. Thanks!!

**Review by matiasgarciaisaia — COMMENTED** [review]
Let's change the `timer:send_interval` thing and merge 👍

We can leave the changes in the GC start time for later.

File: broker/src/scheduler/active_calls_gc.erl [hunk 5]
```diff
@@ -28,7 +28,8 @@ handle_cast(_Msg, State) ->
 
 handle_info(cancel_active_calls, State) ->
   N = verboice_config:minutes_for_cancelling_active_calls(),
-  Count = call_log:cancel_active_calls_for_minutes(N),
+  N_Minutes_Ago = util:seconds_ago(N * 60),
+  Count = call_log:cancel_active_calls_started_before(N_Minutes_Ago),
```
Comment (matiasgarciaisaia) [inline broker/src/scheduler/active_calls_gc.erl:32]: Awesome! 🎉

File: broker/src/scheduler/active_calls_gc.erl [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+handle_cast(_Msg, State) ->
+  {noreply, State}.
+
+handle_info(cancel_active_calls, State) ->
+  N = verboice_config:minutes_for_cancelling_active_calls(),
+  Count = call_log:cancel_active_calls_for_minutes(N),
+  lager:info("GC cancelled ~p calls that stayed active for more than ~p minutes", [Count, N]),
+  Interval = ?MILLISECS_IN_ONE_MINUTE * verboice_config:minutes_between_active_calls_gc_runs(),
+  erlang:send_after(Interval, self(), cancel_active_calls),
```
[the remark this answers is not in the document]
Comment (aeatencio) [reply broker/src/scheduler/active_calls_gc.erl]: Done!

**Comment by aeatencio** [comment]
I think we finished with this, @matiasgarciaisaia. Waiting for the CI to finish and your approval to merge! Thanks a lot!

**Comment by aeatencio** [comment]
Finally, @matiasgarciaisaia, I think we're done. [This](https://github.com/instedd/verboice/pull/912/files#diff-b70a9a4baaeb95364ef8b67a561125bb3ec3330079f50f86db3083ebf09dcd5bR25) is the line I added to avoid the CI crashing 🙂

**Comment by matiasgarciaisaia** [comment]
I've reviewed the PR on Friday but forgot to comment. All is good 👍
